src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -467,5 +471,218 @@ public override T GetResult<T>()
+            }
+
+            /// <summary>
+            /// Encodes the written characters if the character buffer is full.
+            /// </summary>
+            private void EncodeCharactersIfBufferFull()
+            {
+                if (this.charBufferPosition == this.charBuffer.Length)
+                {
+                    this.EncodeCharacters(flushEncoder: false);
+                }
+            }
+
+            /// <summary>
+            /// Encodes characters written so far to a buffer provided by the underyling <see cref="bufferWriter"/>.
+            /// </summary>
+            /// <param name="flushEncoder"><c>true</c> to flush the characters in the encoder; useful when finalizing the output.</param>
+            private void EncodeCharacters(bool flushEncoder)
+            {
+                if (this.charBufferPosition > 0)
+                {
+                    int maxBytesLength = this.Encoding.GetMaxByteCount(this.charBufferPosition);
+                    if (!this.preambleWritten)
+                    {
+                        maxBytesLength += this.encodingPreamble.Length;
+                    }
+
+                    if (this.memory.Length - this.memoryPosition < maxBytesLength)
+                    {
+                        this.CommitBytes();
+                        this.memory = this.bufferWriter.GetMemory(maxBytesLength);
+                    }
+
+                    if (!this.preambleWritten)
+                    {
+                        this.encodingPreamble.Span.CopyTo(this.memory.Span.Slice(this.memoryPosition));
+                        this.memoryPosition += this.encodingPreamble.Length;
+                        this.preambleWritten = true;
+                    }
+
+                    if (MemoryMarshal.TryGetArray(this.memory, out ArraySegment<byte> segment))
+                    {
+                        this.memoryPosition += this.encoder.GetBytes(this.charBuffer, 0, this.charBufferPosition, segment.Array, segment.Offset + this.memoryPosition, flush: flushEncoder);
+                    }
+                    else
+                    {
+                        byte[] rentedByteBuffer = ArrayPool<byte>.Shared.Rent(maxBytesLength);
+                        try
+                        {
+                            int bytesWritten = this.encoder.GetBytes(this.charBuffer, 0, this.charBufferPosition, rentedByteBuffer, 0, flush: flushEncoder);
+                            rentedByteBuffer.CopyTo(this.memory.Span.Slice(this.memoryPosition));
+                            this.memoryPosition += bytesWritten;
+                        }
+                        finally
+                        {
+                            ArrayPool<byte>.Shared.Return(rentedByteBuffer);
+                        }
+                    }
+
+                    this.charBufferPosition = 0;
+
+                    if (this.memoryPosition == this.memory.Length)
+                    {
+                        this.Flush();
```

Comment:
Why is this done recursively instead of iteratively?

Reply:
I don't see recursion or iteration here. The calling method is tasked with encoding `char` to `byte`. When the `byte[]` buffer is full, we need to flush it so there's room for more.